src/Qualify.hs

```diff
@@ -51,19 +51,19 @@ setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [the@(XObj The _ _), t
 setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [def@(XObj Def _ _), sym, expr]) i t) =
   let expr' = setFullyQualifiedSymbols typeEnv globalEnv env expr
   in  XObj (Lst [def, sym, expr']) i t
-setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [letExpr@(XObj Let _ _), bind@(XObj (Arr bindings) bindi bindt), body]) i t) =
-  if even (length bindings)
-  then let Just ii = i
-           lvl = envFunctionNestingLevel env
-           innerEnv = Env Map.empty (Just env) (Just ("let-env-" ++ show (infoIdentifier ii))) [] InternalEnv lvl
-           (innerEnv', bindings') =
-             foldl' (\(e, bs) (s@(XObj (Sym (SymPath _ binderName) _) _ _), o) ->
-                       let qualified = setFullyQualifiedSymbols typeEnv globalEnv e o
-                       in (extendEnv e binderName s, bs ++ [s, qualified]))
-                    (innerEnv, []) (pairwise bindings)
-           newBody = setFullyQualifiedSymbols typeEnv globalEnv innerEnv' body
-       in  XObj (Lst [letExpr, XObj (Arr bindings') bindi bindt, newBody]) i t
-  else XObj (Lst [letExpr, bind, body]) i t -- Leave it untouched for the compiler to find the error.
+setFullyQualifiedSymbols typeEnv globalEnv env (XObj (Lst [letExpr@(XObj Let _ _), bind@(XObj (Arr bindings) bindi bindt), body]) i t) 
+  | odd (length bindings) = XObj (Lst [letExpr, bind, body]) i t -- Leave it untouched for the compiler to find the error.
+  | not (all isSym (evenIndicies bindings)) = XObj (Lst [letExpr, bind, body]) i t -- Leave it untouched for the compiler to find the error.
+  | otherwise = let Just ii = i
```

Comment:
Prior to this change attempting to define `(defn foo [] (let [4 4] true))` in the REPL caused a 'non-exhuastive patterns in lambda' error in `Qualify.hs` and crashed the REPL. This change exchanges that error for a new one: 
```
carp: Can't create let-binder for non-symbol: XObj {obj = Num Int 4.0, info = Just (Info {infoLine = 1, infoColumn = 20, infoFile = "REPL", infoDelete = fromList [], infoIdentifier = 4}), ty = Nothing}
CallStack (from HasCallStack):
  error, called at src/InitialTypes.hs:363:22 in CarpHask-0.2.0.0-5zBtEl5uNipJ0fDU65syPR:InitialTypes
```

Ideally, I think we should try and catch this in the REPL and report the error to the user without crashing, but I'm not exactly sure as to how to accomplish this. It seems like we'd need to put the inner let through another Eval pass? At the moment, if I'm understanding it correctly (and there's a chance I'm not!) I think it passes from Eval straight through Qualify and then to `visit` in `InitialTypes`--or something like that. Any ideas?

Reply:
@hellerve @eriksvedang

Reply:
Hmm, yeah this one was tricky. The `initialTypes` function detects some type errors and have the ability to return TypeError, so perhaps that's where it should be reported? So keep your code but add a proper type error instead of the call to `error` in `extendEnvWithLetBindings`? Or am I missing something?

Reply:
Thanks for the tips! I've added an error to catch this and it seems to work swimmingly, just need a quick double check to make sure the code is free of fishiness (all puns intended).